reposerver/cache/cache.go

```diff
@@ -172,21 +188,98 @@ func (c *Cache) SetGitReferences(repo string, references []*plumbing.Reference)
 	for i := range references {
 		input = append(input, references[i].Strings())
 	}
-	return c.cache.SetItem(gitRefsKey(repo), input, c.revisionCacheExpiration, false)
+	return c.cache.SetItem(gitRefsKey(repo), input, &cacheutil.CacheActionOpts{Expiration: c.revisionCacheExpiration})
 }
 
-// GetGitReferences retrieves resolved Git repository references from cache
-func (c *Cache) GetGitReferences(repo string, references *[]*plumbing.Reference) error {
+// Converts raw cache items to plumbing.Reference objects
+func GitRefCacheItemToReferences(cacheItem [][2]string) *[]*plumbing.Reference {
+	var res []*plumbing.Reference
+	for i := range cacheItem {
+		// Skip empty data
+		if cacheItem[i][0] != "" || cacheItem[i][1] != "" {
+			res = append(res, plumbing.NewReferenceFromStrings(cacheItem[i][0], cacheItem[i][1]))
+		}
+	}
+	return &res
+}
+
+// TryLockGitRefCache attempts to lock the key for the Git repository references if the key doesn't exist, returns the value of
+// GetGitReferences after calling the SET
+func (c *Cache) TryLockGitRefCache(repo string, lockId string, references *[]*plumbing.Reference) (string, error) {
+	// This try set with DisableOverwrite is important for making sure that only one process is able to claim ownership
+	// A normal get + set, or just set would cause ownership to go to whoever the last writer was, and during race conditions
+	// leads to duplicate requests
+	err := c.cache.SetItem(gitRefsKey(repo), [][2]string{{cacheutil.CacheLockedValue, lockId}}, &cacheutil.CacheActionOpts{
+		Expiration:       c.revisionCacheLockTimeout,
+		DisableOverwrite: true})
+	if err != nil {
+		// Log but ignore this error since we'll want to retry, failing to obtain the lock should not throw an error
+		log.Errorf("Error attempting to acquire git references cache lock: %v", err)
+	}
+	return c.GetGitReferences(repo, references)
+}
+
+// Retrieves the cache item for git repo references. Returns foundLockId, error
+func (c *Cache) GetGitReferences(repo string, references *[]*plumbing.Reference) (string, error) {
 	var input [][2]string
-	if err := c.cache.GetItem(gitRefsKey(repo), &input); err != nil {
-		return err
+	err := c.cache.GetItem(gitRefsKey(repo), &input)
+	valueExists := len(input) > 0 && len(input[0]) > 0
+	switch {
+	// Unexpected Error
+	case err != nil && err != ErrCacheMiss:
+		return "", err
```

Comment:
As is this will always log in `util/git/client.go#getRefs` for debug but this is treated as a critical failure so definitely seems like a good idea to have here to make sure it gets logged. Added in an error log for it